### PR TITLE
Implement MSC2730: verifiable forwarded events

### PR DESCRIPTION
**Pull request by tulir**
This adds an implementation of matrix-org/matrix-doc#2730 to Synapse, i.e. it adds a new `PUT /_matrix/client/unstable/net.maunium.msc2730/rooms/{roomId}/event/{eventId}/forward/{targetRoomId}/{txnId}` endpoint and implements validation of incoming events that have the `net.maunium.msc2730` key.

Element web implementation: matrix-org/matrix-js-sdk#1439 / matrix-org/matrix-react-sdk#5117

Signed-off-by: Tulir Asokan &lt;tulir@maunium.net&gt;

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog).
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

**Comment by anoadragon453** [comment]
@tulir is this PR ready for review?

**Comment by tulir** [comment]
I think it is

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -683,6 +684,41 @@ async def _get_events_from_store_or_dest(
...
+
+        room_version = KNOWN_ROOM_VERSIONS.get(room_version_identifier)
+        if not room_version:
+            # We don't support the source room version, so we can't verify the event :(
+            return False, None
+
+        try:
+            source_evt = event_from_pdu_json(source_evt_dict, room_version)
+        except SynapseError:
+            return False, None
```
Comment (richvdh) [inline synapse/handlers/federation.py:710]: as a general rule, eating exceptions like this without giving any clue about what the exception was leads to hard-to-debug failures. I'd recommend logging something before returning.

File: synapse/handlers/federation.py [hunk 2]
```diff
@@ -683,6 +684,41 @@ async def _get_events_from_store_or_dest(
 
         return fetched_events
 
+    _forwarded_key = "net.maunium.msc2730.forwarded"
+
+    async def _validate_forwarded_event(
+        self, event: EventBase
+    ) -> Tuple[bool, Optional[str]]:
```
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/handlers/federation.py:691]: (alternatively: what does it mean if `valid` is False, but an event id is returned? Some docstring would help here).

File: synapse/handlers/federation.py [hunk 3]
```diff
@@ -683,6 +684,41 @@ async def _get_events_from_store_or_dest(
...
+            # We don't support the source room version, so we can't verify the event :(
+            return False, None
+
+        try:
+            source_evt = event_from_pdu_json(source_evt_dict, room_version)
+        except SynapseError:
+            return False, None
+
+        try:
+            checked_evt = await self._check_sigs_and_hash(room_version, source_evt)
```
Comment (richvdh) [inline synapse/handlers/federation.py:713]: I wonder if there is some horrible attack where you can claim that an event is for a different room version than it is, and hence get it to pass the hash checks when it shouldn't...

File: synapse/rest/client/v2_alpha/forward_event.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            if not is_valid_forward:
+                raise SynapseError(
+                    401,
+                    "Event contains invalid forward metadata.",
+                    errcode=self._err_not_forwardable,
+                )
+            # Pass through the old event ID to the new unsigned data
+            event_id = unsigned[self._data_key]["event_id"]
+        elif not has_forward_meta:
+            content[self._data_key] = event_dict
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/forward_event.py:104]: careful: I think this modifies the original event, stored in the cache. You need to copy `content` before modifying it.

**Comment by richvdh** [comment]
Hi @tulir, thanks for this!

It looks generally like a sensible idea, and from the point of view of the MSC process I certainly think there is enough here to demonstrate a workable implementation, but I'm afraid I don't think we can accept this into mainline synapse until the MSC gets wider acceptance, so I'm going to close it for now. We'll be very happy to reopen once the MSC progresses!

**Pull request closed without merging**